bundles-maven-plugin/README.md

```diff
@@ -0,0 +1,230 @@
+  <parent>
+    <artifactId>test.bundles.plugin</artifactId>
+    <groupId>org.apache.test</groupId>
+    <version>1.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>test.app.bundle</artifactId>
+
+  <!-- Packaging is bundle -->
+  <packaging>bundle</packaging>
+  
+  <!-- All dependencies of this module, and all the dependencies of THAT dependency will
+  be included in the produced bundle -->
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.test</groupId>
+      <artifactId>test.app</artifactId>
+      <version>1.0-SNAPSHOT</version>
+    </dependency>
+  </dependencies>
+
+  <build>
+  <!-- OUR PLUGIN REFERENCES -->
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <groupId>org.apache.metron</groupId>
+          <artifactId>bundles-maven-plugin</artifactId>
+          <version>0.4.2</version>
+          <extensions>true</extensions>
+          <configuration>
+          </configuration>
+        </plugin>
+      </plugins>
+    </pluginManagement>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.metron</groupId>
+        <artifactId>bundles-maven-plugin</artifactId>
+        <version>0.4.2</version>
+        <extensions>true</extensions>
+      </plugin>
+    </plugins>
+  </build>
+</project>
+```
+When the module is packaged, it packages all of it's  non-provided dependencies into the bundles /bundled-dependencies directory.
+Thus, to create a bundle of a module's jar and that jar's non-provided dependencies, you add that module to your
+bundle modules dependencies.  You can unzip and examine the bundle in the target directory, and verify 
+it's contents, which should be similar to :
+
+```bash
+-> % tree .
+.
+└── META-INF
+    ├── MANIFEST.MF
+    ├── bundled-dependencies
+    │   ├── log4j-1.2.17.jar
+    │   ├── metron-common-0.4.1.jar
+    │   ├── slf4j-api-1.7.7.jar
+    │   ├── slf4j-log4j12-1.7.7.jar
+    │   └── test.app-1.0-SNAPSHOT.jar
+    └── maven
+        └── org.apache.test
+            └── test.app.bundle
+                ├── pom.properties
+                └── pom.xml
+```
+
+This reflects the testapp project, which has these dependencies :
+
+```xml
+<dependencies>
+    <dependency>
+      <groupId>org.apache.metron</groupId>
+      <artifactId>metron-common</artifactId>
+      <version>0.4.1</version>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>3.8.1</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+```
+metron-common itself is a shaded jar, but it depends on log4j and slf4j, so those libraries are pulled in..
+
+## Quickstart
+
+* Create a new multi module maven project (if you do not have one already)
+* Add a new module for your bundle, it needs only to have a pom.xml
+* Create the pom.xml as above, with the correct plugin and packaging entries, and add dependencies
+for the module you want to bundle.
+* `mvn package`
```

Comment:
Overall to test your PR, I want to do the following... 
  1. Create a bundle
  2. Write some custom code in that bundle
  3. Deploy the bundle somehow and see the code in that bundle execute

I am now trying to complete step 1 and just create a bundle.  Isn't there a `mvn archetype:generate` command that I have to run to create a bundle?  Where are the docs outlining those steps?  I would have expected to see them here, I think, but I don't

Reply:
The getting started section of the plugin documentation tells you how to create a bundle.
The bundles-testing module is a simple result of following those instructions, with an added integration module with an example of using a bundle in code like you are looking to do.

The archetype is for creating specific types of projects that produce artifacts that may include a bundle.  For example an archetype for creating a parser extension that produces an assembly with a bundle in it.

Reply:
I should say the getting started and the quick-start

Reply:
Can you point me to a specific file and section?  I can't find instructions in the 3 new READMEs that I see in the PR.  Maybe I am just missing it somehow.

Reply:
Are you looking for instructions on how to create the bundles-testing project step by step?
like : create a folder.  in that folder create 4 folders.  create a file pom.xml.... ?

Reply:
I had thought that the point of your `bundles-maven-plugin` was to create the files and directory structure of a bundle; ie something like the ` bundles-testing` project.  But based on your answer, I am assuming that with this PR I need to manually create that directory structure and files.  Is that correct?

And then maybe in a future PR you have the code that creates this directory structure auto-magically?

Reply:
In a [previous comment](https://github.com/apache/metron/pull/774#issuecomment-335012079), you have a link to a [`test-plugin-bundles`](https://github.com/ottobackwards/test-bundles-plugin) project.  In that README, you have the following.

> A module generated from the maven quickstart archetype. Note that...

This has made me believe that there is a Maven archetype to create the bundle project.

Reply:
Ok, let me try to detangle this:
- ` A module generated from the maven quickstart archetype` : there is an actual maven quickstart archetype you an use to create a simple maven module project.  This is not related to the plugin:
The expanded version of that comment in the readme would be:

"This module, whose product jar is what we are going to make a bundle of,  was created using the maven quickstart archetype"

I am sorry, I did not understand you were looking at test-plugin-bundles when you made your comment.

Reply:
The point of the bundles maven plugin is to take the product of the module declared in it's dependency section, and create a bundle out of that module's jar, and all of that jar's dependencies.  That bundle having a specific 'directory'/entry structure

Reply:
There is not archetype for creating a bundle setup.  To create a bundle of another module, you create the module folder and a simple pom.xml, listing the target module as a dependency.  As I put in the readme, and tried to show in the sample projects.
Of the sample projects:
test-plugin-bundles: Is a simple project to create a bundle of a project
bundles-testing : is a more complicated but still simple project that has:
- A module for interfaces
- A module for implementation of an interface
- A module that creates a bundle of the implementation module
- An integration test that only references the bundles-lib and the interface module.  This test uses the BundleSystem to load and execute to the implementation

Between these projects and the unit test it is my hope that a reviewer may not only get a handle from example on the basics of bundle projects and loading a bundle in a project, but also back within the agreed upon scope of the review see code examples as well.

Reply:
The Archetype for creating **Extensions**  does more than lay down the folder/pom.xml for creating the bundle.  An Extension may or may not include a bundle, but also includes other things.

Reply:
>  but also back within the agreed upon scope of the review see code examples as well.

Am I going outside some kind of boundaries here?  Please clarify

Reply:
I refer to the discussion of this effort, that this PR's review would be theoretical and based on code review, unit tests, integration tests and the provided test of loading without direct dependencies ( bundles-tests ). 

As stated in the PR description itself: 

> - This review is code review and test code review and running only

Where are we missing each other on this?  It seems we have a disconnect.

Reply:
> This review is code review and test code review and running only

I disagree with that statement.  I don't know why I shouldn't be able to exercise this code like any other PR.  The functionality is here to do that.  You told me that the bundles-testing project has an integration test, that "has load and execute of a bundle."  All I want to do, is do that manually so I can get a feel for the functionality.

But apologies if you'd prefer not to do this.  If you'd like to get someone else's thumbs up on this according to those ground rules, then feel free.

Reply:
I wish you would have stated your disagreement during the community meeting, or in the email list on the 26th when you gave a +1 to this.   

It is unfortunate that we again have gone months on a PR in this now year long effort to no effect, especially since as stated this was supposed to be the 'lighter' review of all the pr's.

That being said, we should not be discouraging reviews under any circumstances, and your review certainly matters.

So please.  Can you explain what would would like to do based on your current understanding now that I have attempted to clarify things, and tried to describe the projects that are present?

Reply:
Nifi does not have this documented, and does not have a 'generic' archetype.  @nickwallen I wonder if an archetype that just creates a simple bundle project would be useful?  Would that be more of what you are looking for than the bundles-testing?  Or would a written guide be the thing?  Does this guide have to document how to create the multi module project?  Does it have to document both existing and from scratch scenarios?

Reply:
Let me clarify, when I say that nifi doesn't have this documented, I mean that their documentation, as mine ( though not included in this PR ) revolves around creating NARs of NiFi components, and not generic NARs.

Reply:
Sorry @ottobackwards.  Not ignoring you.  I have just been busy.  Will get back to this soon.  Thanks for your hard work in helping me through this.

Reply:
no problem, let me know if the new doc is in line with what you are looking for